hydra-cluster/bench/Main.hs

```diff
@@ -114,19 +142,83 @@ main =
+    run options benchDir datasets
 
-  -- TODO(SN): Ideally we would like to say "to re-run use ... " on errors
-  run timeoutSeconds benchDir datasets clusterSize = do
+  run options@Options{timeoutSeconds, clusterSize, startingNodeId} benchDir datasets = do
     putStrLn $ "Test logs available in: " <> (benchDir </> "test.log")
-    withArgs [] . hspec $ bench timeoutSeconds benchDir datasets clusterSize
+    withArgs [] $
+      try (bench startingNodeId timeoutSeconds benchDir datasets clusterSize) >>= \case
+        Left (err :: HUnitFailure) ->
+          benchmarkFailedWith benchDir err
+        Right summary ->
+          benchmarkSucceeded options benchDir summary
 
   saveDataset tmpDir dataset = do
     let txsFile = tmpDir </> "dataset.json"
     putStrLn $ "Writing dataset to: " <> txsFile
     encodeFile txsFile dataset
+
+benchmarkFailedWith :: FilePath -> HUnitFailure -> IO ()
+benchmarkFailedWith benchDir (HUnitFailure _ reason) = do
+  putStrLn $ "Benchmark failed: " <> formatFailureReason reason
+  putStrLn $ "To re-run with same dataset, pass '--work-directory=" <> benchDir <> "' to the executable"
+  exitFailure
+
+benchmarkSucceeded :: Options -> FilePath -> Summary -> IO ()
+benchmarkSucceeded Options{outputDirectory, clusterSize} _ Summary{numberOfTxs, averageConfirmationTime, percentBelow100ms} =
+  maybe dumpToStdout writeMarkdownReportTo outputDirectory
+ where
+  dumpToStdout = do
+    putTextLn $ "Confirmed txs: " <> show numberOfTxs
+    putTextLn $ "Average confirmation time (ms): " <> show averageConfirmationTime
+    putTextLn $ "Confirmed below 100ms: " <> show percentBelow100ms <> "%"
+
+  writeMarkdownReportTo outputDir = do
+    existsDir <- doesDirectoryExist outputDir
+    unless existsDir $ createDirectory outputDir
+    withFile (outputDir </> "end-to-end-benchmarks.md") WriteMode $ \hdl -> do
+      hPut hdl $ encodeUtf8 $ unlines pageHeader
+      hPut hdl $ encodeUtf8 $ unlines $ formattedSummary
+
+  pageHeader :: [Text]
+  pageHeader =
+    [ "--- "
+    , "sidebar_label: 'End-to-End Benchmarks' "
+    , "sidebar_position: 4 "
+    , "--- "
+    , ""
+    , "# End-To-End Benchmark Results "
+    , ""
+    , "This page is intended to collect the latest end-to-end benchmarks \
+      \ results produced by Hydra's Continuous Integration system from \
+      \ the latest `master` code. Please take those results with a grain of \
+      \ salt as they are produced from basic cloud VMs and not controlled \
+      \ hardware. Instead of focusing on the _absolute_ results, the emphasis \
+      \ should be on relative results, eg. how the timings for a scenario \
+      \ evolve as the code changes."
+    , ""
+    , "_Generated at_  " <> show now
+    , ""
+    ]
+
+  {-# NOINLINE now #-}
+  now :: UTCTime
+  now = unsafePerformIO getCurrentTime
+
+  summaryTitle = "Simple Benchmark"
+
+  formattedSummary :: [Text]
+  formattedSummary =
+    [ "## " <> summaryTitle
```

Comment:
Could: make it a bit more clear that this is the best possible result we will ever get

For example by calling this "Trivial transactions" and "nodes running locally without network latency"?

Reply:
I wanted to make that part of the "scenario" by adding an optional `description` field in the `Dataset` record and then storing the dataset in the repository so that we could run in CI a set of "standards" benchmarks without having to hard code the description in the summary page.